core/src/main/java/org/elasticsearch/discovery/zen/fd/NodesFaultDetection.java

```diff
@@ -284,6 +286,18 @@ public void messageReceived(PingRequest request, TransportChannel channel) throw
                     + clusterName + "]");
             }
 
+            final DiscoveryNode currentMasterNode = clusterService.state().nodes().getMasterNode();
+            if (currentMasterNode != null
+                    && request.masterNode != null
+                    && currentMasterNode.equals(request.masterNode) == false) {
+                // this node has a different node as master than the one that sent this ping message, so notify
+                // the sending node that this node does not believe it to be the master
+                // TODO: should we throw a specialized exception type here that can be caught by the
```

Comment:
@bleskes I wonder if we should introduce a new exception type here so that in `NodeFD#run` when it handles exceptions for failed pings, it takes care of this situation and attempt the retries?